### PR TITLE
Stabilizing org.opensearch.cluster.routing.MovePrimaryFirstTests.test…

### LINKED ISSUE #1957: [BUG] Timeout on org.opensearch.cluster.routing.MovePrimaryFirstTests.testClusterGreenAfterPartialRelocation
**Describe the bug**
Caught on PR #1952. The test timed out while waiting for the cluster to become green. Related PR for test: #1445.
```
REPRODUCE WITH: ./gradlew ':server:test' --tests "org.opensearch.cluster.routing.MovePrimaryFirstTests.testClusterGreenAfterPartialRelocation" -Dtests.seed=AF1232B890DC88C7 -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=he-IL -Dtests.timezone=Japan -Druntime.java=17
```

**To Reproduce**
1. Run the above command in the repo. Its a flaky test.

**Expected behavior**
The cluster becomes green and the test does not timeout.

**Plugins**
Core OpenSearch.

**Host/Environment (please complete the following information):**
 - #1952

### PR DESCRIPTION
**Pull request by jainankitk**
…ClusterGreenAfterPartialRelocation

Signed-off-by: Ankit Jain <jain.ankitk@gmail.com>

### Description
The issue is caused due to one of the primary shard being initialized and some replica starts meanwhile. Hence, latch is counted down as half shards are already initialized. Making the check more robust by ensuring no primaries are initializing and not more than 20% of replicas have started on new nodes
 
### Issues Resolved
#1957 
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [] New functionality has been documented.
  - [] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by jainankitk** [comment]
@owaiskazi19 , @andrross - Can you review this PR?

File: server/src/test/java/org/opensearch/cluster/routing/MovePrimaryFirstTests.java [hunk 1]
```diff
@@ -84,19 +86,24 @@ public void testClusterGreenAfterPartialRelocation() throws InterruptedException
...
-                        startedz2n2 = routingNode.numberOfShardsWithState(ShardRoutingState.STARTED);
+                        startedCount += routingNode.numberOfShardsWithState(ShardRoutingState.STARTED);
+                        initz2n2 = routingNode.shardsWithState(ShardRoutingState.INITIALIZING);
                     }
                 }
-                if (startedz2n1 >= primaryShardCount / 2 && startedz2n2 >= primaryShardCount / 2) {
-                    primaryMoveLatch.countDown();
+                if (!Stream.concat(initz2n1.stream(), initz2n2.stream()).anyMatch(s -> s.primary())) {
+                    // All primaries are relocated before 60% of overall shards are started on new nodes
+                    if (primaryShardCount <= startedCount && startedCount <= 6 * primaryShardCount / 5) {
```
Comment (andrross) [inline server/src/test/java/org/opensearch/cluster/routing/MovePrimaryFirstTests.java:104]: Maybe it's obvious, but it is not immediately clear to me why the `6 * primaryShardCount / 5` math is correct for calculating that 60% of shards are started on new nodes. Can you explain how this works?

Comment (jainankitk) [reply server/src/test/java/org/opensearch/cluster/routing/MovePrimaryFirstTests.java:104]: Total number of shards are double the primary shard count - `2 * primaryShardCount`. Hence, 60% of total shards is `3 * total number of shards / 5` which is same as `6 * primaryShardCount / 5`

File: server/src/test/java/org/opensearch/cluster/routing/MovePrimaryFirstTests.java [hunk 2]
```diff
@@ -113,6 +120,6 @@ public void testClusterGreenAfterPartialRelocation() throws InterruptedException
             internalCluster().stopRandomNode(InternalTestCluster.nameFilter(z1n1));
             internalCluster().stopRandomNode(InternalTestCluster.nameFilter(z1n2));
```
Comment (owaiskazi19) [inline server/src/test/java/org/opensearch/cluster/routing/MovePrimaryFirstTests.java:121]: Do we have to shutdown nodes z2n1 and z2n2 as well here?

Comment (jainankitk) [reply server/src/test/java/org/opensearch/cluster/routing/MovePrimaryFirstTests.java:121]: There are 4 nodes in the cluster. If we shutdown all 4, cluster will not be green. We want to shutdown all excluded nodes (in this case 2) after 60% of total shards have relocated to z2n1 and z2n2. Due to [#1445 ] all primaries would have started in those 60% and hence, cluster will become eventually green

File: server/src/test/java/org/opensearch/cluster/routing/MovePrimaryFirstTests.java (hunk 1 above)
Comment (andrross) [reply server/src/test/java/org/opensearch/cluster/routing/MovePrimaryFirstTests.java:104]: Got it. I suggest creating an intermediate variable just to make this more readable, like:

```
final int totalShardCount = primaryShardCount * 2;
if (primaryShardCount <= startedCount && startedCount <= totalShardCount * 3 / 5) {
```

Comment (jainankitk) [reply server/src/test/java/org/opensearch/cluster/routing/MovePrimaryFirstTests.java]: Sure, makes sense

**Review by andrross — APPROVED** [review]

**Comment by jainankitk** [comment]
Gradle check failure did not reproduce locally:

```
akjain@88665a3f79e7 OpenSearch % ./gradlew ':server:test' --tests "org.opensearch.cluster.routing.MovePrimaryFirstTests.testClusterGreenAfterPartialRelocation" -Dtests.seed=EFD233475A0C0A16 -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=sr-ME -Dtests.timezone=Indian/Reunion -Druntime.java=17 > /tmp/log
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
OpenJDK 64-Bit Server VM warning: Ignoring option --illegal-access=warn; support was removed in 17.0
WARNING: A terminally deprecated method in java.lang.System has been called
WARNING: System::setSecurityManager has been called by org.opensearch.bootstrap.BootstrapForTesting (file:/Users/akjain/ws/OpenSearch/test/framework/build/distributions/framework-2.0.0-SNAPSHOT.jar)
WARNING: Please consider reporting this to the maintainers of org.opensearch.bootstrap.BootstrapForTesting
WARNING: System::setSecurityManager will be removed in a future release
WARNING: A terminally deprecated method in java.lang.System has been called
WARNING: System::setSecurityManager has been called by org.gradle.api.internal.tasks.testing.worker.TestWorker (file:/Users/akjain/.gradle/wrapper/dists/gradle-7.3.3-all/4295vidhdd9hd3gbjyw1xqxpo/gradle-7.3.3/lib/plugins/gradle-testing-base-7.3.3.jar)
WARNING: Please consider reporting this to the maintainers of org.gradle.api.internal.tasks.testing.worker.TestWorker
WARNING: System::setSecurityManager will be removed in a future release
akjain@88665a3f79e7 OpenSearch %
```

**Comment by jainankitk** [comment]
Restart gradle check

**Comment by owaiskazi19** [comment]
start gradle check

**Review by owaiskazi19 — APPROVED** [review]

**Comment by andrross** [comment]
The fact that [gradle check 2275](https://ci.opensearch.org/logs/ci/workflow/OpenSearch_CI/PR_Checks/Gradle_Check/gradle_check_2275.log) failed test MovePrimaryFirstTests.testClusterGreenAfterPartialRelocation suggests that this PR doesn't actually fix the issue with that flaky test, right?

**Comment by jainankitk** [comment]
> The fact that [gradle check 2275](https://ci.opensearch.org/logs/ci/workflow/OpenSearch_CI/PR_Checks/Gradle_Check/gradle_check_2275.log) failed test MovePrimaryFirstTests.testClusterGreenAfterPartialRelocation suggests that this PR doesn't actually fix the issue with that flaky test, right?

I ran it locally with the fix for 100 times and it passed 98 times. Even 2 times that it failed looked unrelated reasons
Screen Shot 2022-02-07 at 7 37 10 PM

**Comment by owaiskazi19** [comment]
@jainankitk may be try adding a timeout of 120 sec for ensureGreen() method so that nodes get available within the same time. Not sure if it will solve the above issue but let's give it a try.

**Comment by jainankitk** [comment]
> @jainankitk may be try adding a timeout of 120 sec for ensureGreen() method so that nodes get available within the same time. Not sure if it will solve the above issue but let's give it a try.

The 2 times that test failed with the fix did not seem anything to do with the timeout

**Pull request merged**